sdk/src/main/java/io/dapr/actors/runtime/DaprStateAsyncProvider.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT License.
+ */
+
+package io.dapr.actors.runtime;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import reactor.core.publisher.Mono;
+
+import java.io.IOException;
+import java.util.Collection;
+
+/**
+ * State Provider to interact with Dapr runtime to handle state.
+ */
+class DaprStateAsyncProvider {
+
+  /**
+   * Shared Json serializer/deserializer as per Jackson's documentation, used only for this class.
+   */
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+  private final AppToDaprAsyncClient daprAsyncClient;
+
+  private final ActorStateProviderSerializer serializer;
+
+  DaprStateAsyncProvider(AppToDaprAsyncClient daprAsyncClient, ActorStateProviderSerializer serializer) {
+    this.daprAsyncClient = daprAsyncClient;
+    this.serializer = serializer;
+  }
+
+  <T> Mono<T> load(String actorType, String actorId, String stateName, Class<T> clazz) {
+    Mono<String> result = this.daprAsyncClient.getState(actorType, actorId, stateName);
+
+    return result.map(s -> {
+      if (s == null) {
+        return (T)null;
+      }
+
+      try {
+        return this.serializer.deserialize(s, clazz);
+      } catch (IOException e) {
+        throw new RuntimeException(e);

```

Comment:
Throw more meaningful custom DaprException with error code.

Reply:
We need to refactor all exception cases. It will be handled as a separate change.